src/sage/groups/finitely_presented.py

```diff
@@ -1532,7 +1601,169 @@ def alexander_matrix(self, im_gens=None):
+        r"""
+        Return the characteristic varieties of the group ``self``. There are several definitions
+        of the characteristic varieties of a group `G`, see e.g. [CS1999a]_. Let `\Lambda` be the
+        group algebra of `G/G'` and `\mathbb{T}` its associated algebraic variety (a torus). Each
+        element `\xi\in\mathbb{T}` defines a local system of coefficients and the `k` th-characteristic
+        variety is
+
+        .. MATH::
+
+            V_k(G) = \{\xi\in\mathbb{T}\mid \dim H^1(G;\xi)\geq k\}
+
+        These varieties are defined by ideals in `\Lambda`
+
+        INPUT:
+
+        - ``ring`` -- base ring (default: ``QQ``). The base ring of the
+          group algebra
+
+        - ``matrix_ideal`` -- optional. An abelian Alexander matrix and an ideal,
+          to avoid recomputations.
+
+        - ``abelianized`` -- optional. The data of the abelianization.
+
+        - ``groebner`` -- boolean (default: ``False``). If set to
+          ``True`` the minimal associated primes of the ideals and their
+          groebner bases are computed-. It has no effect if the base ring
+          is not a field.
+
+        OUTPUT:
+
+        If ``groebner`` is ``False`` a list of ideals defining the characteristic varieties.
+        If it is True, a list of lists for groebner bases for each ideal.
+
+        EXAMPLES::
+
+            sage: L = [2*(i, j) + 2* (-i, -j) for i, j in ((1, 2), (2, 3), (3, 1))]
+            sage: G=FreeGroup(3) / L
+            sage: G.char_var(groebner=True)
+            [[(f1 - 1, f2 - 1, f3 - 1),
+             (f1 + 1, f2 - 1, f3 - 1),
+             (f1 - 1, f2 - 1, f3 + 1),
+             (f3^2 + 1, f1 - f3, f2 - f3),
+             (f1 - 1, f2 + 1, f3 - 1)],
+             [(f1 - 1, f2 - 1, f3 - 1),
+             (f1*f3 + 1, f2 - 1),
+             (f1*f2 + 1, f3 - 1),
+             (f2*f3 + 1, f1 - 1),
+             (f2*f3 + 1, f1 - f2),
+             (f2*f3 + 1, f1 - f3),
+             (f1*f3 + 1, f2 - f3)]]
+        """
+        if matrix_ideal is None:
+            A, R, ideal = self.abelian_alexander_matrix(ring=ring, abelianized=abelianized, simplified=True)
+        else:
+            A, ideal = matrix_ideal
+            R = A.base_ring()
+        res = []
+        S = R.polynomial_ring()
+        ideal = [S(_) for _ in ideal]
+        for j in range(1, A.ncols()):
+            L = [p.polynomial_construction()[0] for p in A.minors(j)]
+            J = R.ideal(L + ideal)
+            res.append(J)
+        if not groebner or not R.base_ring().is_field():
+            return res
+        if R.ngens() == 1:
+            res0 = [[S(_) for _ in J.gens()] for J in res]
+            res1 = []
+            for JL in res0:
+                J = S.ideal(JL)
+                LJ = J.minimal_associated_primes()
+                fct = [id.groebner_basis() for id in LJ]
+                fct = [a for a in fct if S.gen() not in a]
+                if fct != [(S.one(),)]:
```

Comment:
Mirco-optimization: pull out the `[(S.one(),)]` into a variable to avoid reconstructing the list and tuple every loop.